### PR TITLE
fix: introducing feature flagged meilisearch

### PR DESCRIPTION
**Pull request by rebelchris**
Running meilisearch as a option beyond feature flags. (Would say just for internal testing for now)
Not fully optimized, but leveraging all things we already have in place 🙏

### CODE REVIEW

File: src/schema/search.ts [hunk 1]
```diff
@@ -228,6 +246,29 @@ export const resolvers: IResolvers<unknown, Context> = traceResolvers({
       ctx,
       info,
     ): Promise<ConnectionRelay<GQLPost> & { query: string }> => {
+      const gbClient = getUserGrowthBookInstace(ctx.userId);
```
Comment (idoshamun) [inline src/schema/search.ts:249]: Our server side feature flag doesn't support the team attribute which will force us to use hardcoded user ids and bloat the boot endpoint. This is something we want to consider here. Can we perhaps send a search version from the client to `searchPosts` and `searchPostSuggestions` like we do with the feed? Will be much easier I believe

**Review by idoshamun — CHANGES_REQUESTED** [review]
Someone is really keen to have new search 😉 
Overall looks great I had a suggestion and also `searchPostSuggestions` implementation is missing. We need to check their autocomplete support

File: src/schema/search.ts [hunk 2]
```diff
@@ -201,9 +224,21 @@ export const resolvers: IResolvers<unknown, Context> = traceResolvers({
       getSession(ctx.userId, id),
     searchPostSuggestions: async (
       source,
-      { query }: { query: string },
+      { query, version }: { query: string; version: number },
       ctx,
     ): Promise<GQLSearchPostSuggestionsResults> => {
+      if (version === 2) {
+        const hits = await searchMeili(
+          `q=${query}&attributesToRetrieve=post_id,title`,
```
Comment (idoshamun) [inline src/schema/search.ts:232]: It's fine for now but pretty sure they have something dedicated for auto complete

**Review by idoshamun — APPROVED** [review]

File: src/schema/search.ts (hunk 2 above)
Comment (rebelchris) [reply src/schema/search.ts:232]: Yeah they do, not sure if we just want to try it for now and see what we want to use?
They have a million and one options basically 😅

Just keeping it lean to evaluate for now (seems a tad slow still)

Comment (idoshamun) [reply src/schema/search.ts:232]: No worries let's get it started and optimize later

**Comment by rebelchris** [comment]
Deploying for test drive.
Feel free to still leave feedback

File: .infra/Pulumi.prod.yaml [hunk 3]
```diff
@@ -62,6 +62,12 @@ config:
       secure: AAABAAAcNFaMFyerA4UVMKYeQYf69uQ0mMfl2I0DIiKOBa6HwtbnTXoA1N4OwidUgnIRlGzdYrgQWzQTMsimcE/Yarpceg==
     magniOrigin:
       secure: AAABAHxH1dQ0e4nHwnseCd65qMd7qz3vSsKV/H0vMQIKWjxS4H7BemSLMpPGEH7YsDLUM9hEMw==
+    meiliIndex:
+      secure: AAABADypa9PiyaCl1GQdpBI+5tcJRA0G894+ZL/lqfW3iWEetKlYQg==
```
Comment (capJavert) [inline .infra/Pulumi.prod.yaml:66]: I think we can manage indexes as enums right? If we have more in the future and they are not exactly secret?

File: src/integrations/meilisearch.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  hits: Meili[];
+}
+export const searchMeili = async (params: string): Promise<Meili[]> => {
+  const headers = new Headers({
+    Authorization: `Bearer ${meiliToken}`,
+  });
+  const requestOptions = {
+    method: 'GET',
+    headers,
+  };
```
Comment (capJavert) [inline src/integrations/meilisearch.ts:23]: I would advise using `POST` endpoint in the future, it has a lot more options and capabilities. https://www.meilisearch.com/docs/reference/api/search#search-in-an-index-with-post

File: src/schema/search.ts [hunk 5]
```diff
@@ -224,10 +259,27 @@ export const resolvers: IResolvers<unknown, Context> = traceResolvers({
...
     ): Promise<ConnectionRelay<GQLPost> & { query: string }> => {
+      if (args.version === 2) {
+        const meilieSearchRes = await searchMeili(
+          `q=${args.query}&attributesToRetrieve=post_id`,
+        );
+
+        const meilieArgs: FeedArgs & { ids: string[] } = {
+          ...args,
+          ids: meilieSearchRes.map((x) => x.post_id),
+        };
```
Comment (capJavert) [inline src/schema/search.ts:274]: What fields are available in our documents? Also, we should also have dev only API key and index so we can connect to it and use in development locally.

**Review by capJavert — COMMENTED** [review]
Awesome work on bringing it online fast 🚀

File: .infra/Pulumi.prod.yaml (hunk 3 above)
Comment (rebelchris) [reply .infra/Pulumi.prod.yaml:66]: Potentially yes, for now we only have one just added one more later of security 😂

File: src/integrations/meilisearch.ts (hunk 4 above)
Comment (rebelchris) [reply src/integrations/meilisearch.ts:23]: Nice both seem to have many options already 🤯
Lets see how poc goes if good we can switch 👍

File: src/schema/search.ts (hunk 5 above)
Comment (rebelchris) [reply src/schema/search.ts:274]: Not many think top of head
- time
- post_id
- title
- content 

In this query we only need id as we do internal feed resolver anyway 🙏

**Pull request merged**